### PR TITLE
cr me daddy

### PR DESCRIPTION
**Pull request by hadar759**
Think i'm done - prob some things you aren't really ok with, some names (at least 1 that i can easily think off) that need to be changed. Anyways, yeah it works and it's pretty ok cr it now pwease :3

### CODE REVIEW

File: DiscordBot/Bot.cs [hunk 1]
```diff
@@ -36,6 +37,7 @@ public async Task Run()
         {
             // We received a message
             DiscordClient.MessageCreated += MessageHandler;
+            DiscordClient.MessageCreated += HandleTicTacToeScreen;
```
Comment (tomgrin10) [inline DiscordBot/Bot.cs]: `DiscordBot` doesn't need to know about TicTacToe's existence

File: DiscordBot/Commands/PingCommand.cs [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 namespace DiscordBot.Commands
 {
-    class Ping: ICommand
+    class PingCommand: ICommand
```
Comment (tomgrin10) [inline DiscordBot/Commands/PingCommand.cs:8]: 🔥

File: DiscordBot/Bot.cs [hunk 3]
```diff
@@ -64,5 +66,30 @@ private async Task MessageHandler(MessageCreateEventArgs e)
             
             await command.Run(e.Message);
         }
+
+        public async Task HandleTicTacToeScreen(MessageCreateEventArgs e)
```
Comment (tomgrin10) [inline DiscordBot/Bot.cs]: > `DiscordBot` doesn't need to know about TicTacToe's existence

File: DiscordBot/Commands/ICommandGetter.cs [hunk 4]
```diff
@@ -6,6 +6,6 @@ namespace DiscordBot.Commands
 {
     interface ICommandGetter
```
Comment (tomgrin10) [inline DiscordBot/Commands/ICommandGetter.cs]: I thought of a better name for the class: `CommandProvider` :)

File: DiscordBot/Commands/TicTacToeCommand.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DSharpPlus.Entities;
+using Microsoft.CSharp.RuntimeBinder;
+using TicTacToe;
+using static TicTacToe.SquareState;
```
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs]: Don't do `using static`, it introduces confusing code.
Not recommended in general.

File: DiscordBot/Commands/TicTacToeCommand.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+using static TicTacToe.GameState;
```
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs]: Same

File: DiscordBot/Commands/TicTacToeCommand.cs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+                case "stop":
+                    await msg.RespondAsync(StopText(msg));
+                    break;
+
+                case "help":
+                    await msg.RespondAsync(HelpText());
+                    break;
+
+                default:
```
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs:51]: Export the body of this block to a method

Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs:51]: I don't really understand what you do in here, explain later

File: DiscordBot/Commands/TicTacToeCommand.cs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        public TicTacToeCommand(TicTacToeManager manager)
+        {
+            Manager = manager;
+        }
+
+        public async Task Run(DiscordMessage msg)
+        {
+            var splitMsg = CommandParser.Split(msg.Content);
+
+            switch (splitMsg[1])
```
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs:55]: Really like the usage of methods `StartText`, `StopText`...
But make it so the methods send the text themselves, not return it, and rename them to `RunStartSubcommand`, `RunStopSubcommand`...
Essentially make them 'handlers' for the commands

File: DiscordBot/Commands/TicTacToeCommand.cs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        }
+
+        private string HelpText() =>
+            "Each turn the bot will send a tic tac toe screen." +
+            " Place an :x: wherever you want to play and send the screen." +
+            "\nMake sure to avoid spaces and other unnecessary characters." +
+            "\nFirst to complete a row, column, or diagonal wins.";
+
+        /// <summary>
+        /// The correct text to send after the user tried to start the game
```
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs]: Not really what the method does though is it?
It also actually starts the game, by talking to the manager.

Also, the method description can't be a noun, the method is not **The correct text**, but **Returns the correct text**

**Review by tomgrin10 — CHANGES_REQUESTED** [review]

File: DiscordBot/Commands/TicTacToeCommand.cs [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            {
+                TicTacToeGame game = Manager.GetGame(msg);
+                char[,] screen = game.ScreenToMatrix(translationDict);
+
+                return MatrixToString(screen);
+            }
+        }
+
+        /// <summary>
+        /// The correct text to send after the user tried to stop the game
```
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs]: Same as `StartText`

File: DiscordBot/Commands/TicTacToeCommand.cs [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+            " Place an :x: wherever you want to play and send the screen." +
+            "\nMake sure to avoid spaces and other unnecessary characters." +
+            "\nFirst to complete a row, column, or diagonal wins.";
+
+        /// <summary>
+        /// The correct text to send after the user tried to start the game
+        /// </summary>
+        /// <param name="msg"></param>
+        /// <returns></returns>
+        private string StartText(DiscordMessage msg)
```
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs]: Are you starting the text?

File: DiscordBot/Program.cs [hunk 12]
```diff
@@ -1,21 +1,29 @@
...
-            { "ping", new Ping() }
+            { "ping", new PingCommand() },
+            {"tic", _ticTacToeCommand},
+            {"tac", _ticTacToeCommand},
+            {"toe", _ticTacToeCommand},
         };
         
         static void Main(string[] args)
         {
             // fuck tom.... <3
```
Comment (tomgrin10) [inline DiscordBot/Program.cs]: 🤬

File: DiscordBot/Program.cs [hunk 13]
```diff
@@ -1,21 +1,29 @@
...
-            var configGetter = new JsonConfigGetter($"config.json");
+            var path = Path.GetDirectoryName(Path.GetDirectoryName(Path.GetDirectoryName(System.IO.Directory.GetCurrentDirectory())));
```
Comment (tomgrin10) [inline DiscordBot/Program.cs]: Why......

Just do `var path = "config.json"`

File: TicTacToe/TicTacToeGame.cs [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+using Microsoft.CSharp.RuntimeBinder;
+using static TicTacToe.SquareState;
+using static TicTacToe.GameState;
+
+namespace TicTacToe
+{
+    public class TicTacToeGame
+    {
+        public SquareState[,] Screen { get; set; }
+        // public SquareState MySign { get; set; }
```
Comment (tomgrin10) [inline TicTacToe/TicTacToeGame.cs]: Delete commented lines

File: TicTacToe/TicTacToeGame.cs [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+        }
+        
+        /// <summary>
+        /// Does one game turn
+        /// </summary>
+        /// <param name="screenMatrix">The user inputted screen</param>
+        /// <param name="translationDict">A dictionary to translate the inputted screen into a game screen</param>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        public GameState DoTurn<T>(T[,] screenMatrix, Dictionary<T, SquareState> translationDict)
```
Comment (tomgrin10) [inline TicTacToe/TicTacToeGame.cs]: Is there a reason to receive translationDict every turn? Maybe receive it in the constructor?

File: TicTacToe/TicTacToeGame.cs [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+        /// Returns whether the game has resulted in a draw
+        /// </summary>
+        /// <returns></returns>
+        private bool IsDraw() => GetAvailableMoves().Count == 0;
+        
+        /// <summary>
+        /// Returns which sign won the game
+        /// </summary>
+        /// <returns>X, O, or Empty if no one won</returns>
+        private SquareState WinningSign()
```
Comment (tomgrin10) [inline TicTacToe/TicTacToeGame.cs]: Bad method name

File: DiscordBot/Commands/TicTacToeCommand.cs (hunk 7 above)
Comment (tomgrin10) [inline DiscordBot/Commands/TicTacToeCommand.cs]: The `default` block should say:
```
await RunTurn(msg);
```

**Review by tomgrin10 — CHANGES_REQUESTED** [review]

File: DiscordBot/Commands/ICommandGetter.cs (hunk 4 above)
Comment (hadar759) [reply DiscordBot/Commands/ICommandGetter.cs]: wow!

File: DiscordBot/Commands/TicTacToeCommand.cs (hunk 5 above)
Comment (hadar759) [reply DiscordBot/Commands/TicTacToeCommand.cs]: ok

File: DiscordBot/Commands/TicTacToeCommand.cs (hunk 6 above)
Comment (hadar759) [reply DiscordBot/Commands/TicTacToeCommand.cs]: ok

File: DiscordBot/Commands/TicTacToeCommand.cs (hunk 8 above)
Comment (hadar759) [reply DiscordBot/Commands/TicTacToeCommand.cs:55]: u hate me cause you ain't me :D 🥇

File: DiscordBot/Commands/TicTacToeCommand.cs (hunk 11 above)
Comment (hadar759) [reply DiscordBot/Commands/TicTacToeCommand.cs]: 👎

Comment (hadar759) [reply DiscordBot/Commands/TicTacToeCommand.cs]: virgin

File: DiscordBot/Program.cs (hunk 13 above)
Comment (hadar759) [reply DiscordBot/Program.cs]: If i remember correctly, it didn't work :(

File: TicTacToe/TicTacToeGame.cs (hunk 14 above)
Comment (hadar759) [reply TicTacToe/TicTacToeGame.cs]: ok

File: TicTacToe/TicTacToeGame.cs (hunk 16 above)
Comment (hadar759) [reply TicTacToe/TicTacToeGame.cs]: jealous bitch

File: DiscordBot/Commands/TicTacToeCommand.cs (hunk 7 above)
Comment (hadar759) [reply DiscordBot/Commands/TicTacToeCommand.cs]: Like, without checking if the game exists? Or you just wanted me to make all the methods send the message themselves?

File: TicTacToe/TicTacToeGame.cs (hunk 16 above)
Comment (hadar759) [reply TicTacToe/TicTacToeGame.cs]: GetWhichSignWon() is ok? I don't really have an idea what to rename it lol

File: DiscordBot/Program.cs (hunk 13 above)
Comment (hadar759) [reply DiscordBot/Program.cs]: I'll try to change it tomorrow I guess when my computer will actually be running at an acceptable speed and I'd be able to turn on the bot easily ♥

File: DiscordBot/Program.cs (hunk 12 above)
Comment (hadar759) [reply DiscordBot/Program.cs]: poopoo head

File: DiscordBot/Commands/PingCommand.cs (hunk 2 above)
Comment (hadar759) [reply DiscordBot/Commands/PingCommand.cs:8]: Why thank you!

File: DiscordBot/Bot.cs (hunk 1 above)
Comment (hadar759) [reply DiscordBot/Bot.cs]: I'll deal with it tomorrow lol

File: DiscordBot/Program.cs (hunk 13 above)
Comment (hadar759) [reply DiscordBot/Program.cs]: nah doesn't work lol

Comment (hadar759) [reply DiscordBot/Program.cs]: ez

**Pull request merged**